org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/JDTLanguageServer.java

```diff
@@ -176,6 +181,25 @@ public void didChangeConfiguration(DidChangeConfigurationParams params) {
 			Preferences prefs = Preferences.createFrom(javaConfig);
 			preferenceManager.update(prefs);
 		}
+		if (!preferenceManager.getPreferences().isJavaFormatEnabled()) {
+			Unregistration unregistration = new Unregistration(Preferences.FORMATTING_ID, Preferences.TEXT_DOCUMENT_FORMATTING);
```

Comment:
should check if dynamic registration of formatting capabilities is available.

in ClientPreferences, new method should check if v3Supported and capabilities.getTextDocument().getFormatting().getDynamicRegistration() is true